content/docs/faq-ajax.md

```diff
@@ -31,35 +31,43 @@ The example API returns a JSON object like this:
 
 ```jsx
 class MyComponent extends React.Component {
-  state = {
-    error: null,
-    isLoaded: false,
-    items: []
-  };
+  constructor(props) {
+    super(props);
+    this.state = {
+      error: null,
+      isLoaded: false,
+      items: []
+    };
+  }
 
   componentDidMount() {
     fetch("https://api.example.com/items")
       .then(res => res.json())
-      .then(result =>
-        this.setState({
-          isLoaded: true,
-          items: result.items
-        })
+      .then(
+        (result) => {
+          this.setState({
+            isLoaded: true,
+            items: result.items
+          })
+        },
+        // Note: it's important to handle errors here
+        // instead of a catch() block so that we don't swallow
+        // exceptions from actual bugs in components.
```

Comment:
IMO this needs more explanation - most people do not know the second argument to  `.then` is an error handler.

Very good point though, did not think of that.

Reply:
Happy to take as a further PR :-)

Reply:
@gaearon, would you mind explaining why exceptions would be swallowed when using `catch`? From my understanding of how promises work, it would seem that `catch` would actually catch the error, but if there's an error in the error callback passed (lines 56-61), then _that_ would be swallowed.

Reply:
My point was that

```js
something
  .then(() => this.setState(...))
  .catch(() => this.setState(...))
```

is very easy to mess up because usually people try to `catch` exceptions from `something` and don't realize exception in the first `setState` would *also* trigger second `setState`. If your component throws *rendering*, by that point you have nothing to `setState` on because the tree has been unmounted. So the original error often ends up ignored and not logged. React 16 made this better by always logging rendering errors (even when caught) but it's still a problem.

Reply:
Ahh, I see! Then perhaps something like this could be clearer?

```jsx
    fetch("https://api.example.com/items")
      .then(res => res.json())
      .then(
        (result) => {
          this.setState({
            isLoaded: true,
            items: result.items
          })
        },
        (error) => {
          this.setState({
            isLoaded: true,
            error
          })        
        }
      )
      .catch(
        (error) => {
          // If your component throws an exception in `render`
          // it has to be caught here.
        }
      )
```

Or this:

```jsx
    fetch("https://api.example.com/items")
      .then(res => res.json())
      .then(
        (result) => {
          this.setState({
            isLoaded: true,
            items: result.items
          })
        }
      )
      .catch(
        (error) => {
          this.setState({
            isLoaded: true,
            error
          })        
        }
      )
      .catch(
        (error) => {
          // If your component throws an exception in `render`
          // from the previous `catch` `setState` call it has
          // to be handled here.
        }
      )
```

Because I believe the way it is right now, if the `setState` in line 57 causes an exception in the component, that exception will be swallowed.

I wouldn't mind opening a PR, I know you're likely super busy with other stuff.

Reply:
Here's a sandbox to play around with this: https://codesandbox.io/s/7zw59n0130

Reply:
>if the setState in line 57 causes an exception in the component, that exception will be swallowed

It will be an uncaught rejection, wouldn't it? That's a different thing from being swallowed. Uncaught rejections are reported by Chrome (and throw in Node).

Reply:
@migueloller I opened an issue to discuss this topic further with a bit more visibility (outside of this closed PR) - care to take the discussion over here: https://github.com/reactjs/reactjs.org/issues/309?

Reply:
@alexkrolick, thanks! That's a good idea 😄

Reply:
I vote for adding this discussion as a read more link in this particular doc page.